packages/lib-classifier/src/plugins/tasks/survey/components/SurveyTask.js

```diff
@@ -1,24 +1,91 @@
 import { Box } from 'grommet'
+import { observer } from 'mobx-react'
 import PropTypes from 'prop-types'
-import React from 'react'
+import React, { useEffect } from 'react'
 
 import Choice from './components/Choice'
 import Chooser from './components/Chooser'
 
-function SurveyTask({
-  answers = {},
+export function SurveyTaskContainer ({
+  annotation,
+  task,
   disabled = false,
-  filters = {},
-  previousChoice = '',
-  handleAnswers = () => {},
-  handleChoice = () => {},
-  handleDelete = () => {},
-  handleFilter = () => {},
-  handleIdentify = () => {},
-  selectedChoice = '',
-  selectedChoiceIds = [],
-  task
 }) {
+  const [answers, setAnswers] = React.useState({})
+  const [filters, setFilters] = React.useState({})
+  const [previousChoice, setPreviousChoice] = React.useState('')
+  const [selectedChoice, setSelectedChoice] = React.useState('')
+
+  useEffect(() => {
+    return () => {
+      if (annotation._choiceInProgress) {
+        annotation.setChoiceInProgress(false)
+      }
+    }
+  })
+
+  function handleAnswers (answers) {
+    setAnswers(answers)
+  }
+
+  function handleChoice (selectedChoice) {
+    if (selectedChoice === '') {
+      annotation.setChoiceInProgress(false)
+    } else {
+      annotation.setChoiceInProgress(true)
+    }
+
+    if (annotation?.value?.map(item => item.choice).includes(selectedChoice)) {
+      const existingAnnotationValue = annotation?.value?.find(value => value.choice === selectedChoice)
+      setAnswers(existingAnnotationValue.answers)
+      setSelectedChoice(selectedChoice)
+    } else {
+      setAnswers({})
+      setSelectedChoice(selectedChoice)
+    }
+  }
+
+  function handleDelete (deletedChoice) {
+    if (annotation?.value?.map(item => item.choice).includes(deletedChoice)) {
+      const value = annotation?.value?.filter(item => item.choice !== deletedChoice) || []
+      annotation.update(value)
+    }
+    
+    setAnswers({})
+    setPreviousChoice(deletedChoice)
+    setSelectedChoice('')
+    
+    annotation.setChoiceInProgress(false)
+  }
+
+  function handleFilter (characteristicId, valueId) {
+    let newFilters = Object.assign({}, filters)
+    if (valueId) {
+      newFilters[characteristicId] = valueId
+    } else if (characteristicId) {
+      delete newFilters[characteristicId]
+    } else {
+      newFilters = {}
+    }
+
+    setFilters(newFilters)
+  }
+
+  function handleIdentify () {
+    const parsedFilters = JSON.parse(JSON.stringify(filters))
```

Comment:
Seems kind of slow to repeatedly parse the filters if they've already been parsed. Why does this convert them to a string then parse them again for each identification?

Reply:
Great, question, I'm not sure. I tried searching for a past explanation, and the best I can tell is that I copied it [from PFE](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/survey/index.jsx#L110) and maybe it was creating a copy of the `filters` object for saving to the annotation value, though I'm not sure why. However, `answers` works similarly without any such copying, and I tried filtering on different characteristics, making multiple identifications, changing filters, and I see no reason for the JSON methods (everything works as expected without the JSON methods).

I've removed the `JSON.parse` and `JSON.stringify` in 35fbca34a3523776ef20c6fc40d28312ec63e5ad.

Reply:
Deja vu!
https://github.com/zooniverse/Panoptes-Front-End/pull/5076/files#r234709160

Reply:
You’re right, it’s cloning the `filters` object.

Reply:
If the `filters` prop does need to be cloned, that can be done during the render, wrapped in a `useMemo` hook to memo-ise the copied object.

Reply:
I can't think of how the `filters` object would have a value that references an object, a filter characteristic can only be only be one thing, i.e. color can only be red, so I think the deep clone (with JSON methods) is not necessary.

After testing things out, now I also don't think copying the `filters` object is necessary at all. I created a workflow with two survey tasks - https://local.zooniverse.org:8080/?project=1634&workflow=3719 - and I don't see any issues with filters and answers from one task to the other and the annotations look as expected.

So I think it's ok not to copy the `filters` object before adding it to the `annotation.value` (how code is currently), but please let me know if you can think of anything I'm missing.

Reply:
Seems good to me, and we can catch any changes with tests that check the annotation value after an identification.